cmd/login.go

```diff
@@ -11,40 +13,73 @@ import (
 	"golang.org/x/crypto/ssh/terminal"
 )
 
-func init() {
-	var (
-		loginCmd = &cobra.Command{
-			Use:   "login",
-			Short: "Login to AppLand",
-			Args:  cobra.MaximumNArgs(1),
-			Run: func(cmd *cobra.Command, args []string) {
-				reader := bufio.NewReader(os.Stdin)
-				context, err := config.GetCurrentContext()
-				if err != nil {
-					fail(err)
-				}
+type PasswordReader func() ([]byte, error)
+
+var stdinPasswordReader = func() ([]byte, error) {
+	return terminal.ReadPassword(0)
+}
+
+func NewLoginCommand(connecter Connecter, stdin io.Reader, passwordReader PasswordReader) *cobra.Command {
+	return &cobra.Command{
+		Use:   "login",
+		Short: "Login to AppLand",
+		Args:  cobra.MaximumNArgs(1),
+		PersistentPreRun: func(cmd *cobra.Command, args []string) {
+			config.LoadCLIConfig()
+			// defer connecting to the server until Run (below)
+		},
+		Run: func(cmd *cobra.Command, args []string) {
+			reader := bufio.NewReader(stdin)
 
-				fmt.Printf("logging into %s\n\n", context.GetURL())
-				fmt.Printf("login: ")
-				login, err := reader.ReadString('\n')
+			api := connecter()
+
+			context, err := config.GetCurrentContext()
+			if err != nil {
+				fail(err)
+			}
+
+			fmt.Printf("logging into %s\n\n", context.GetURL())
+			fmt.Printf("login: ")
+			login, err := reader.ReadString('\n')
+			if err != nil {
+				fail(err)
+			}
+			login = strings.TrimSpace(login)
+
+			validAPIKey := false
+			data, err := base64.RawURLEncoding.DecodeString(login)
+			if err == nil && len(strings.Split(string(data), ".")) == 2 {
```

Comment:
What if the user has a dot in the name?

Reply:
If their username is valid Bas64, and when decoded it has a`.` in it, we'll guess wrong and login will fail.

Reply:
I agree, check for “at least one dot”.

Reply:
Not sure I understand the scenario you're trying to protect against....

This code doesn't look for a dot in the login. It looks for a dot in the decoded base64 string. So, `alan@app.land` doesn't get flagged as an API key, because decoding it fails.